ansible_kernel/kernel.py

```diff
@@ -765,44 +846,81 @@ def get_module_doc(self, module):
+            logger.info("NO")
+        return self.runner_thread.is_alive()
 
-    def do_shutdown(self, restart):
+    def cancel_callback(self):
+        logger.info('called')
+        return self.shutdown_requested
 
-        logger = logging.getLogger('ansible_kernel.kernel.do_shutdown')
+    def finished_callback(self, runner):
+        logger.info('called')
+        self.shutdown = True
+        if not self.shutdown_requested:
+            self.queue.put(StatusMessage(['PlaybookEnded', {}]))
 
-        if self.ansible_process is None:
-            logger.debug("No ansible process")
-            return
+    def do_shutdown(self, restart):
 
-        try:
-            current_process = psutil.Process(self.ansible_process.pid)
-            children = current_process.children(recursive=True)
-            for child in children:
-                print('Child pid is {}'.format(child.pid))
-        except psutil.NoSuchProcess:
-            pass
 
         if self.is_ansible_alive():
-            logger.info('killing ansible {0}'.format(self.ansible_process.pid))
-            try:
-                self.ansible_process.kill()
-            except psutil.NoSuchProcess:
-                logger.info('process already dead {0}'.format(self.ansible_process.pid))
-            for child in children:
-                logger.info('killing ansible sub {0}'.format(child.pid))
-                try:
-                    child.kill()
-                except psutil.NoSuchProcess:
-                    logger.info('process already dead {0}'.format(child.pid))
-
-        logger.info('stopping helper')
-        self.helper.stop()
-        self.helper = None
-        self.tasks_counter = 0
+            self.shutdown = False
+            self.shutdown_requested = True
+
+            while not self.shutdown:
+                if not self.is_ansible_alive():
+                    break
+                logger.info("waiting for shutdown")
+                time.sleep(1)
+            logger.info("shutdown complete")
+
+        self.shutdown_requested = False
+        self.runner_thread = None
+        self.runner = None
+        if self.helper is not None:
+            self.helper.stop()
+            self.helper = None
 
-        logger.info('clean up')
-        self.ansible_process = None
         return {'status': 'ok', 'restart': restart}
+
+    def _format_output(self, result):
+        if 'stdout_lines' in result:
+            return '\n'.join(result['stdout_lines'])
+        return ""
+
+    def _format_error(self, result):
+        if 'stderr_lines' in result:
+            return '\n'.join(result['stderr_lines'])
+        return ""
+
+    def _dump_results(self, result):
+
+        r = result.copy()
+        for key in ['_ansible_verbose_always',
+                    '_ansible_no_log',
+                    '_ansible_parsed',
+                    'invocation']:
+            if key in r:
+                del r[key]
+        if 'stdout' in r:
+            if r['stdout']:
```

Comment:
Is this because the kernel itself is going to produce its own output relative to what's going on?

Reply:
This is because stdout looks terrible inside JSON formatted data.